### PR TITLE
Global err control

### PR DESCRIPTION
**Pull request by phlptp**
### Summary

If merged this pull request will add improved error handling at global and local level,  introduce global and local error calls. 

### Proposed changes

- add functions for global and local errors
- add error propagation rules to the CommonCore and CoreBroker
- update the error handling in Federate State
- some checks

### CODE REVIEW

**Comment by phlptp** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/helics/core/TimeCoordinator.cpp  6
- src/helics/core/FederateState.cpp  3
- tests/helics/system_tests/ErrorTests.cpp  27
         

Clones removed
==============
+ src/helics/core/CoreBroker.cpp  -11
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/GMLC-TDC/HELICS/pullRequest?prid=5114566&bid=16815164)

**Comment by phlptp** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/helics/core/TimeCoordinator.cpp  6
- tests/helics/system_tests/ErrorTests.cpp  25
- src/helics/core/FederateState.cpp  3
         

Clones removed
==============
+ src/helics/core/CoreBroker.cpp  -11
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/GMLC-TDC/HELICS/pullRequest?prid=5114566&bid=16815164)

**Comment by phlptp** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/helics/core/TimeCoordinator.cpp  6
- src/helics/core/FederateState.cpp  3
- tests/helics/system_tests/ErrorTests.cpp  25
         

Clones removed
==============
+ src/helics/core/CoreBroker.cpp  -11
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/GMLC-TDC/HELICS/pullRequest?prid=5114566&bid=16815164)

**Comment by phlptp** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- tests/helics/system_tests/ErrorTests.cpp  27
- src/helics/core/TimeCoordinator.cpp  6
- src/helics/core/FederateState.cpp  3
         

Clones removed
==============
+ src/helics/core/CoreBroker.cpp  -11
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/GMLC-TDC/HELICS/pullRequest?prid=5114566&bid=16815164)

**Comment by phlptp** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/helics/core/TimeCoordinator.cpp  6
- src/helics/core/FederateState.cpp  3
- tests/helics/system_tests/ErrorTests.cpp  27
         

Clones removed
==============
+ src/helics/core/CoreBroker.cpp  -11
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/GMLC-TDC/HELICS/pullRequest?prid=5114566&bid=16815164)

**Comment by phlptp** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/helics/core/FederateState.cpp  3
- tests/helics/system_tests/ErrorTests.cpp  27
- src/helics/core/TimeCoordinator.cpp  6
         

Clones removed
==============
+ src/helics/core/CoreBroker.cpp  -11
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/GMLC-TDC/HELICS/pullRequest?prid=5114566&bid=16815164)

**Comment by phlptp** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/helics/core/TimeCoordinator.cpp  6
- tests/helics/system_tests/ErrorTests.cpp  27
- tests/helics/application_api/FederateTests.cpp  2
- src/helics/core/FederateState.cpp  3
         

Clones removed
==============
+ src/helics/core/CoreBroker.cpp  -11
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/GMLC-TDC/HELICS/pullRequest?prid=5114566&bid=16815164)

**Comment by phlptp** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/helics/core/TimeCoordinator.cpp  6
- tests/helics/system_tests/ErrorTests.cpp  27
- src/helics/core/FederateState.cpp  3
- tests/helics/application_api/FederateTests.cpp  2
         

Clones removed
==============
+ src/helics/core/CoreBroker.cpp  -11
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/GMLC-TDC/HELICS/pullRequest?prid=5114566&bid=16815164)

**Comment by phlptp** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/helics/core/FederateState.cpp  3
- src/helics/core/TimeCoordinator.cpp  6
- tests/helics/application_api/FederateTests.cpp  2
- tests/helics/system_tests/ErrorTests.cpp  27
         

Clones removed
==============
+ src/helics/core/CoreBroker.cpp  -11
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/GMLC-TDC/HELICS/pullRequest?prid=5114566&bid=16815164)

**Comment by phlptp** [comment]
Not at present, though that is going to need to be a capability to handle dynamic federations, so this is a small step in that direction.  I would imagine some cases where a failed init could be corrected or a federate could join later.    There are a few scenarios this brings up that I don't think are fully resolved yet but I think what is in place will handle the desired use cases for the moment.  Once we stat dealing with more dynamic federations this will need to get a lot more rigorous.

File: src/helics/cpp98/Federate.hpp [hunk 1]
```diff
@@ -502,6 +502,24 @@ class Federate {
         helicsFederateSetGlobal(fed, valueName.c_str(), value.c_str(), hThrowOnError());
     }
 
+    /** generate a local federate error
+    @param error_code an error code to give to the error
+    @param error_string a string message associated wit the error
```
Comment (nightlark) [inline src/helics/cpp98/Federate.hpp]:
```suggestion
    @param error_string a string message associated with the error
```

File: src/helics/cpp98/Federate.hpp [hunk 2]
```diff
@@ -502,6 +502,24 @@ class Federate {
...
+    */
+    void localError(int error_code, const std::string &error_string)
+    {
+        helicsFederateLocalError(fed, error_code, error_string.c_str());
+    }
+
+    /** generate a global error to terminate the federation
+    @param error_code an error code to give to the error
+    @param error_string a string message associated wit the error
```
Comment (nightlark) [inline src/helics/cpp98/Federate.hpp]:
```suggestion
    @param error_string a string message associated with the error
```

**Comment by phlptp** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/helics/core/FederateState.cpp  3
- src/helics/core/TimeCoordinator.cpp  6
- tests/helics/system_tests/ErrorTests.cpp  27
- tests/helics/application_api/FederateTests.cpp  2
         

Clones removed
==============
+ src/helics/core/CoreBroker.cpp  -11
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/GMLC-TDC/HELICS/pullRequest?prid=5114566&bid=16815164)

**Comment by phlptp** [comment]
I am going to let the recent changes percolate through the CI systems and see if any other tweaks are needed, and check it over again merge the recent changes before merging maybe later tonight.

**Comment by phlptp** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- tests/helics/system_tests/ErrorTests.cpp  27
- src/helics/core/FederateState.cpp  3
- tests/helics/application_api/FederateTests.cpp  2
- src/helics/core/TimeCoordinator.cpp  6
         

Clones removed
==============
+ src/helics/core/CoreBroker.cpp  -11
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/GMLC-TDC/HELICS/pullRequest?prid=5114566&bid=16815164)

**Comment by phlptp** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- tests/helics/system_tests/ErrorTests.cpp  27
- tests/helics/application_api/FederateTests.cpp  2
- src/helics/core/FederateState.cpp  3
- src/helics/core/TimeCoordinator.cpp  6
         

Clones removed
==============
+ src/helics/core/CoreBroker.cpp  -11
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/GMLC-TDC/HELICS/pullRequest?prid=5114566&bid=16815164)

**Pull request merged**